plugins/src/main/java/tech/pegasys/pantheon/plugins/internal/PantheonPluginContextImpl.java

```diff
@@ -0,0 +1,170 @@
+
+  private enum Lifecycle {
+    UNINITIALIZED,
+    REGISTERING,
+    REGISTERED,
+    STARTING,
+    STARTED,
+    STOPPING,
+    STOPPED
+  }
+
+  private Lifecycle state = Lifecycle.UNINITIALIZED;
+  private final Map<Class<?>, ? super Object> serviceRegistry = new HashMap<>();
+  private final List<PantheonPlugin> plugins = new ArrayList<>();
+
+  public <T> void addService(final Class<T> serviceType, final T service) {
+    checkArgument(serviceType.isInterface(), "Services must be Java interfaces.");
+    checkArgument(
+        serviceType.isInstance(service),
+        "The service registered with a type must implement that type");
+    serviceRegistry.put(serviceType, service);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> Optional<T> getService(final Class<T> serviceType) {
+    return Optional.ofNullable((T) serviceRegistry.get(serviceType));
+  }
+
+  public void registerPlugins(final Path pluginsDir) {
+    checkState(state == Lifecycle.UNINITIALIZED, "PantheonContext has already been registered.");
+    if (pluginsDir == null) {
+      LOG.debug("Plugins are disabled.");
+      return;
+    }
+    state = Lifecycle.REGISTERING;
+
+    LOG.debug("Searching for plugins in {}", pluginsDir.toAbsolutePath().toString());
+    try (final Stream<Path> pluginFilesList = Files.list(pluginsDir)) {
+      final URL[] pluginJarURLs =
+          pluginFilesList
+              .filter(p -> p.getFileName().toString().endsWith(".jar"))
+              .map(PantheonPluginContextImpl::pathToURIOrNull)
+              .toArray(URL[]::new);
+      final ServiceLoader<PantheonPlugin> serviceLoader =
+          ServiceLoader.load(
+              PantheonPlugin.class,
+              new URLClassLoader(pluginJarURLs, this.getClass().getClassLoader()));
+
+      for (final PantheonPlugin plugin : serviceLoader) {
+        try {
+          plugin.register(this);
+          LOG.debug("Registered plugin of type {}.", plugin.getClass().getName());
+        } catch (final Exception e) {
+          LOG.error(
+              "Error registering plugin of type {}, start and stop will not be called. \n{}",
+              plugin.getClass(),
+              e);
+          continue;
+        }
+        plugins.add(plugin);
+      }
+
+    } catch (final MalformedURLException e) {
+      LOG.error("Error converting files to URLs, could not load plugins", e);
+    } catch (final IOException e) {
+      LOG.error("Error enumerating plugins, could not load plugins", e);
+    }
+    LOG.debug("Plugin registration complete.");
+    state = Lifecycle.REGISTERED;
+  }
+
+  public void startPlugins() {
+    checkState(
+        state == Lifecycle.REGISTERED,
+        "PantheonContext should be in state %s but it was in %s",
+        Lifecycle.REGISTERED,
+        state);
+    state = Lifecycle.STARTING;
+    final Iterator<PantheonPlugin> pluginsIterator = plugins.iterator();
+
+    while (pluginsIterator.hasNext()) {
```

Comment:
Do we really want to start plugins sequentially ?

Reply:
The other option being parallel?  I don't see parallel startup ending well.